Index: src/uk/me/parabola/mkgmap/osmstyle/actions/NotContainedFilter.java
===================================================================
--- src/uk/me/parabola/mkgmap/osmstyle/actions/NotContainedFilter.java	(revision 0)
+++ src/uk/me/parabola/mkgmap/osmstyle/actions/NotContainedFilter.java	(revision 0)
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2015.
+ * 
+ * This program is free software; you can redistribute it and/or modify it under the terms
+ * of the GNU General Public License version 3 or version 2 as published by the Free
+ * Software Foundation.
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ */
+package uk.me.parabola.mkgmap.osmstyle.actions;
+
+import java.util.regex.Pattern;
+
+import uk.me.parabola.mkgmap.reader.osm.Element;
+import uk.me.parabola.mkgmap.reader.osm.TagDict;
+import uk.me.parabola.mkgmap.scan.SyntaxException;
+
+/**
+ * This can be used to filter out redundant values.<br>
+ * <br>
+ * The filter checks whether the value is contained within another tag's value. If so, a
+ * null string is returned.<br>
+ * <br>
+ * Another tag should consist of values separated by a delimiter (semicolon ';' by
+ * default).<br>
+ * <br>
+ * Example: after<br>
+ * <code>set tag='Aa#Bb#Cc';</code><br>
+ * <code>set tag2='Ax|not-contained:#:tag';</code> would set the tag 'tag2' to the value
+ * 'Ax', whereas<br>
+ * <code>set tag3='Aa|not-contained:#:tag';</code> would set leave the tag 'tag3'
+ * undefined.
+ *
+ * @author Maxim Duester
+ */
+public class NotContainedFilter extends ValueFilter {
+	private String separator;
+	private short tagKey;
+
+	public NotContainedFilter(String arg) {
+		String[] temp = arg.split(":");
+
+		if (temp.length < 2)
+			throw new SyntaxException(
+					"Missing tag to compare in style not-contained command: " + arg);
+
+		// set the separator (default to ;)
+		if (temp[0].length() > 0)
+			separator = temp[0];
+		else
+			separator = ";";
+		// set the tag short value
+		tagKey = TagDict.getInstance().xlate(temp[1]);
+	}
+
+	public String doFilter(String value, Element el) {
+		if (value == null)
+			return null;
+
+		String tagValue = el.getTag(tagKey);
+		if (tagValue == null)
+			return value;
+
+		// split uses a regex we need to replace special characters
+		String[] temp = tagValue.split(Pattern.quote(separator));
+
+		for (String s : temp)
+			if (s.equals(value))
+				return null;
+
+		// nothing found => value not in tag's value
+		return value;
+	}
+}
Index: src/uk/me/parabola/mkgmap/osmstyle/actions/PartFilter.java
===================================================================
--- src/uk/me/parabola/mkgmap/osmstyle/actions/PartFilter.java	(revision 3413)
+++ src/uk/me/parabola/mkgmap/osmstyle/actions/PartFilter.java	(working copy)
@@ -35,7 +35,7 @@
  * part:#:-1 returns Ee
  * part:#:2  returns Bb
  * part:#:-2 returns Dd
- * part:#>1  returns Bb#Cc#Dd#Ee#
+ * part:#>1  returns Bb#Cc#Dd#Ee
  * part:#<5  returns Aa#Bb#Cc#Dd#
  * part:#<-1 returns Aa#Bb#Cc#Dd#
  *
Index: src/uk/me/parabola/mkgmap/osmstyle/actions/ValueBuilder.java
===================================================================
--- src/uk/me/parabola/mkgmap/osmstyle/actions/ValueBuilder.java	(revision 3413)
+++ src/uk/me/parabola/mkgmap/osmstyle/actions/ValueBuilder.java	(working copy)
@@ -236,6 +236,9 @@
 		case "country-ISO":
 			item.addFilter(new CountryISOFilter());
 			break;
+		case "not-contained":
+			item.addFilter(new NotContainedFilter(arg));
+			break;
 		default:
 			throw new SyntaxException(String.format("Unknown filter '%s'", cmd));
 		}
Index: src/uk/me/parabola/mkgmap/osmstyle/actions/ValueItem.java
===================================================================
--- src/uk/me/parabola/mkgmap/osmstyle/actions/ValueItem.java	(revision 3413)
+++ src/uk/me/parabola/mkgmap/osmstyle/actions/ValueItem.java	(working copy)
@@ -47,7 +47,7 @@
 			Element e = tagname_is_local ? local_el : el;
 			String tagval = e.getTag(tagKey);
 			if (filter != null)
-				value = filter.filter(tagval,el);
+				value = filter.filter(tagval, local_el);
 			else
 				value = tagval;
 		}
